### ISSUE TITLE
Add support for numpy structured array conversion to and from

### ISSUE THREAD

**Issue by deanm0000** [pola-rs/polars#8564]
### Problem description

A while back there was [this](https://stackoverflow.com/questions/75913916/how-to-create-a-useful-polars-dataframe-from-a-numpy-structured-array) question which would have benefited from being able to convert a structured array into a polars df.

Another usage for the to_structured_array use case is feeding it to something like statsmodels so that names and dtypes are preserved.

Right now I can do something like 

```
from numpy.lib.recfunctions import unstructured_to_structured
x_cols=['col1','col2','col3']
X=unstructured_to_structured(df.select(x_cols).to_numpy())
X.dtype.names=x_cols
```
but then the dtypes are all floats. I don't think that matters in the context of OLS but in other contexts it might be more important to retain dtypes.

Alternatively, other packages like patsy and formulaic produce their own Matrix datatype which is just an np.ndarray but it retains the column names in a way that statsmodels recognizes. I haven't dug into the source code enough to see how that works but that would be enough for me although perhaps not as extensible.

**Comment by s-banach** [comment]
Can you use `df.to_pandas()`?

**Comment by deanm0000** [comment]
In a literal sense, yes, but I'm trying to reduce dependence on pandas.

**Comment by tikkanz** [comment]
Does the answer to [this stackoverflow question](https://stackoverflow.com/questions/72054388/how-to-read-a-c-struct-or-numpy-record-array-into-a-polars-dataframe) help?

**Comment by zundertj** [comment]
It seems, for Statmodels + formulaic interface, rather easy to create your own wrapper, see [working with large datasets in the user guide](https://www.statsmodels.org/stable/large_data.html#subsetting-your-data). Something along the lines of (note: not tested):

```python
class DataSet(dict):
    def __init__(self, df):
        self._df = df

    def __getitem__(self, key):
        try:
            return df[key].to_numpy()
        except:
            raise KeyError
```

**Comment by deanm0000** [comment]
@tikkanz yes it's helpful although their return trip method uses the structured array that they started from. That being said we can fix that like this:

```
npdf=np.empty(df.height, [(x, df.head(0).get_column(x).to_numpy().dtype.str) for x in df.columns])
for x in df.columns:
    npdf[x]=df.get_column(x).to_numpy()
```

**Comment by alexander-beedie** [comment]
Done; we'll have native init/export support for numpy structured/record arrays in the upcoming `0.17.12` release 👍

**Comment by deanm0000** [comment]
What's the syntax to export a structured array? Also, thanks for your efforts, they're much appreciated.

**Comment by alexander-beedie** [comment]
> What's the syntax to export a structured array?

Have a look at the pull request - it's all detailed there ;)
TLDR: `df.to_numpy( structured=True )`
